File: DesignPattern/other_patterns/reader_writer_lock_pattern.py
import threading
import time

class ReaderWriterLock:
    def __init__(self):
        self.lock = threading.Lock()
        self.resource_lock = threading.Lock()
        self.resource = 0
        self.readers = 0

    def read(self, reader_id):
        with self.lock:
            self.readers += 1
            if self.readers == 1:
                # First reader acquires resource lock
                self.resource_lock.acquire()

        # Reading data
        print(f"Reader {reader_id} reads: {self.resource}")

        with self.lock:
            self.readers -= 1
            if self.readers == 0:
                # Last reader releases resource lock
                self.resource_lock.release()

    def write(self, writer_id, new_data):
        with self.lock:
            # Acquire resource lock
            self.resource_lock.acquire()

            # Writing data
            self.resource = new_data
            print(f"Writer {writer_id} writes: {self.resource}")

            # Release resource lock
            self.resource_lock.release()

def reader_thread(reader_id):
    for _ in range(5):
        rw_lock.read(reader_id)
        time.sleep(0.1)  # Simulate some work

def writer_thread(writer_id):
    for i in range(5):
        new_data = writer_id * 100 + i
        rw_lock.write(writer_id, new_data)
        time.sleep(0.1)  # Simulate some work

rw_lock = ReaderWriterLock()

reader_threads = [threading.Thread(target=reader_thread, args=(i,)) for i in range(3)]
writer_threads = [threading.Thread(target=writer_thread, args=(i,)) for i in range(2)]

for thread in reader_threads + writer_threads:
    thread.start()

for thread in reader_threads + writer_threads:
    thread.join()



'''
- Python provides a built-in module called threading for working with threads. However, Python's Global Interpreter Lock (GIL) can limit true parallel execution 
of threads in some cases. For scenarios where you need to implement a Reader-Writer Lock pattern in Python, you can use the threading module to manage concurrent 
access. Here's a basic example of a Reader-Writer Lock pattern in Python:

- In this Python example: We create a ReaderWriterLock class to manage the concurrent access of readers and writers to a shared resource. It uses 
Python's built-in threading.Lock for synchronization.

- Readers acquire a lock to access the shared resource, while writers acquire both the main lock and a resource-specific lock.

- The reader_thread and writer_thread functions simulate reader and writer threads, respectively. Readers read the shared resource, and writers write new data to it.

- We create and start three reader threads and two writer threads. The readers and writers run concurrently, and the lock implementation ensures that 
multiple readers can read simultaneously, while only one writer can write at a time.

- This example demonstrates a basic Reader-Writer Lock pattern in Python using the threading module. Keep in mind that Python's Global Interpreter Lock (GIL) 
may limit the true parallel execution of threads, so for CPU-bound tasks, you might consider using multiprocessing or alternative concurrency 
mechanisms for greater performance.
'''

File: DesignPattern/other_patterns/reader_writer_lock_pattern.cpp
#include <iostream>
#include <thread>
#include <shared_mutex>

std::shared_mutex rwLock;  // Create a shared reader-writer lock
int sharedData = 0;

void reader(int id) {
    std::this_thread::sleep_for(std::chrono::milliseconds(100));  // Simulate some work
    std::shared_lock<std::shared_mutex> lock(rwLock);
    std::cout << "Reader " << id << " reads: " << sharedData << std::endl;
}

void writer(int id) {
    std::this_thread::sleep_for(std::chrono::milliseconds(100));  // Simulate some work
    std::unique_lock<std::shared_mutex> lock(rwLock);
    sharedData = id;
    std::cout << "Writer " << id << " writes: " << sharedData << std::endl;
}

int main() {
    std::thread writers[3];
    std::thread readers[5];

    for (int i = 0; i < 3; ++i) {
        writers[i] = std::thread(writer, i);
    }

    for (int i = 0; i < 5; ++i) {
        readers[i] = std::thread(reader, i);
    }

    for (int i = 0; i < 3; ++i) {
        writers[i].join();
    }

    for (int i = 0; i < 5; ++i) {
        readers[i].join();
    }

    return 0;
}


/*
- The Reader-Writer Lock pattern is a synchronization pattern that allows multiple readers to access a resource simultaneously but ensures that only 
one writer can access the resource at a time. This pattern is used to manage concurrent access to shared resources while optimizing performance. 
In C++, you can implement a Reader-Writer Lock using the std::shared_mutex introduced in C++17. Here's an example:

- In this example: We use the std::shared_mutex to create a shared reader-writer lock, rwLock. The std::shared_lock is used by readers to lock the resource 
in a shared (read) mode, and the std::unique_lock is used by writers to lock the resource exclusively.

- The reader function simulates a reader. It sleeps for some time to simulate work, then acquires a shared lock using std::shared_lock to read the shared data and prints it.

- The writer function simulates a writer. It sleeps for some time to simulate work, then acquires a unique lock using std::unique_lock to write data to 
the shared resource and prints it.

- In the main function, we create and start three writer threads and five reader threads.

- The writers and readers run concurrently, and you'll notice that multiple readers can access the shared data simultaneously, while only one writer 
can access the data at a time.

- This example demonstrates the Reader-Writer Lock pattern in C++ using the std::shared_mutex. It provides efficient synchronization for shared data 
access while allowing multiple readers to read simultaneously and ensuring exclusive access for writers.

*/
